### PR TITLE
ddl: add auto random && shard_row_id_bits compatibility for temporary table

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: see https://github.com/pingcap/tidb/pull/24168

### What is changed and how it works?

Proposal: [temporary table](https://github.com/pingcap/tidb/issues/24169) 

What's Changed:  add auto random && shard_row_id_bits compatibility for temporary table

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

File: planner/core/preprocess.go [hunk 1]
```diff
@@ -616,7 +624,7 @@ func (p *preprocessor) checkCreateTableGrammar(stmt *ast.CreateTableStmt) {
 			p.err = err
 			return
 		}
-		isPrimary, err := checkColumnOptions(colDef.Options)
+		isPrimary, err := checkColumnOptions(stmt.TemporaryKeyword != ast.TemporaryNone, colDef.Options)
```
Comment (bb7133) [inline planner/core/preprocess.go:627]:
```
tidb> create temporary table t(a bigint key auto_random);
ERROR 8006 (HY000): `auto_random` is unsupported on temporary tables.
```
The current PR forbids 'auto_random' with 'MySQL/local temporary' table(although it is a no-op keyword when `tidb_enable_noop_functions=true`).

Please confirm if it is expected(I think it should be fine), if yes, please add a test case for it.

**Review by djshow832 — COMMENTED** [review]
How about `ALTER TABLE`?
`ALTER TABLE t SHARD_ROW_ID_BITS = 4;`

**Review by djshow832 — APPROVED** [review]

File: planner/core/preprocess.go [hunk 2]
```diff
@@ -895,7 +907,31 @@ func (p *preprocessor) checkAlterTableGrammar(stmt *ast.AlterTableStmt) {
...
+	tempTableType := tableInfo.Meta().TempTableType
 	for _, spec := range specs {
+		if tempTableType != model.TempTableNone {
+			for _, option := range spec.Options {
+				if option.Tp == ast.TableOptionShardRowID {
+					p.err = ErrOptOnTemporaryTable.GenWithStackByArgs("shard_row_id_bits")
+					return
+				}
+			}
+		}
```
Comment (djshow832) [inline planner/core/preprocess.go:933]: Put it into the `switch case` below with `ast.AlterTableOption`.

Comment (djshow832) [reply planner/core/preprocess.go:933]: Besides, I think the code will be much simplified if you put it into `ddl.AlterTable`.

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**